src/Agent.Service/Windows/AgentListener.cs

```diff
@@ -0,0 +1,276 @@
+            // since SCM cache the ServiceHost in memory, sometime we need update the servicehost as well, in this way we can upgrade the ServiceHost as well.
+
+            DirectoryInfo dirInfo = new DirectoryInfo(GetDiagnosticFolderPath());
+            FileInfo[] updateLogs = dirInfo.GetFiles("SelfUpdate-*-*.log.*") ?? new FileInfo[0];
+            if (updateLogs.Length == 0)
+            {
+                // totally wrong, we are not even get a update log.
+                return AgentUpdateResult.Failed;
+            }
+            else
+            {
+                String latestLogFile = null;
+                DateTime latestLogTimestamp = DateTime.MinValue;
+                foreach (var logFile in updateLogs)
+                {
+                    int timestampStartIndex = logFile.Name.IndexOf("-") + 1;
+                    int timestampEndIndex = logFile.Name.LastIndexOf(".log") - 1;
+                    string timestamp = logFile.Name.Substring(timestampStartIndex, timestampEndIndex - timestampStartIndex + 1);
+                    DateTime updateTime;
+                    if (DateTime.TryParseExact(timestamp, "yyyyMMdd-HHmmss", null, DateTimeStyles.None, out updateTime) &&
+                        updateTime > latestLogTimestamp)
+                    {
+                        latestLogFile = logFile.Name;
+                        latestLogTimestamp = updateTime;
+                    }
+                }
+
+                if (string.IsNullOrEmpty(latestLogFile) || latestLogTimestamp == DateTime.MinValue)
+                {
+                    // we can't find update log with expected naming convention.
+                    return AgentUpdateResult.Failed;
+                }
+
+                if (DateTime.UtcNow - latestLogTimestamp > TimeSpan.FromSeconds(15))
+                {
+                    // the latest update log we find is more than 15 sec old, the update process is busted.
+                    return AgentUpdateResult.Failed;
+                }
+                else
+                {
+                    string resultString = Path.GetExtension(latestLogFile).TrimStart('.');
+                    AgentUpdateResult result;
+                    if (Enum.TryParse<AgentUpdateResult>(resultString, true, out result))
+                    {
+                        // return the result indicated by the update log.
+                        return result;
+                    }
+                    else
+                    {
+                        // can't convert the result string, return failed to stop the service.
+                        return AgentUpdateResult.Failed;
+                    }
+                }
+            }
+        }
+
+        private string GetDiagnosticFolderPath()
+        {
+            return Path.Combine(Path.GetDirectoryName(Path.GetDirectoryName(Assembly.GetEntryAssembly().Location)), "_diag");
+        }
+
+        private Process CreateAgentListener()
```

Comment:
I would like to be able to distinguish in the Agent.Listener.exe process whether the agent was invoked: from agent service (Win svc) or from auto-logon (and by elimination, infer invocation must be interactively from the command prompt).